pyatv/protocols/companion/api.py

```diff
@@ -282,6 +290,59 @@ async def hid_command(self, down: bool, command: HidCommand) -> None:
             "_hidC", {"_hBtS": 1 if down else 2, "_hidC": command.value}
         )
 
+    async def hid_event(self, x: int, y: int, mode: HidEventMode) -> None:
+        """Send a HID command."""
+        x = max(x, 0)
+        y = max(y, 0)
+        x = min(x, int(TOUCHPAD_WIDTH))
+        y = min(y, int(TOUCHPAD_HEIGHT))
+        await self._send_event(
+            identifier="_hidT",
+            content={"_ns": (time.time_ns()-self._base_timestamp), "_tFg": 1, "_cx": x, "_tPh": mode.value, "_cy": y}
+        )
+
+    async def touch(self, start_x: int, start_y: int, end_x: int, end_y: int, duration_ms: int):
+        """ Generate a touch gesture from start to end x,y coordinates (in range [0,1000])
+        in a given time (in milliseconds)
+        :param start_x: Start x coordinate
+        :param start_y: Start y coordinate
+        :param end_x: End x coordinate
+        :param end_y: Endi x coordinate
+        :param duration_ms: Time in milliseconds to reach the end coordinates
+        """
+        end_time = time.time_ns() + duration_ms * 1000000
+        x: float = float(start_x)
+        y: float = float(start_y)
+        await self.hid_event(int(x), int(y), HidEventMode.Press)
+        sleep_time = float(TOUCHPAD_DELAY_MS / 1000)
+        current_time = time.time_ns()
+        while current_time < end_time:
```

Comment:
nitpick, but I wonder if using a simple for loop would make more sense here:
Calculate the total amount of iterations needed by dividing the duration_ms the delay or something of that nature.

Reply:
No because you cannot be sure how long it will take to send the events (depending on client device activity and cpu skills) and the timestamp sent has to be very precise (in nanoseconds). But maybe this is not a big deal not to be very precise with this... in that way we are sure to finish after the exact given duration and coordinates

Reply:
Well, from what I can guess at,
the _ns flag records the actual timestamp when the touch event was recorded on the client. This is to mitigate network lag. The Apple TV looks at the _ns field to determine when the touch actually occurred (relative the previous touches). So _ns can be precalculated and doesn't depend on the actual time when the packet is sent.

Reply:
I am not so sure about that : we are talking about 10ms which is very low and may be affected by cpu charge and affect the movement. 
You have to consider that this library may be used on low end device. 
I am not opposed to simplification but this has to be tested and confirmed

Reply:
I believe @thiccaxe is correct here. It makes the most sense that the client includes the *actual* timestamp of the event. In reality, it should not make that much of a difference other than possibly some stuttering.